dipy/reconst/multi_voxel.py

```diff
@@ -72,22 +110,61 @@ def new_fit(self, data, *, mask=None, **kwargs):
                 data_to_fit[ii : ii + vox_per_chunk]
                 for ii in range(0, data_to_fit.shape[0], vox_per_chunk)
             ]
+
+            # func_kwargs : dict or sequence, optional
+            # Keyword arguments to `func` or sequence of keyword arguments
+            # to `func`: one item for each item in the input list.
+            kwargs_chunks = []
+            for ii in range(0, data_to_fit.shape[0], vox_per_chunk):
+                kw = kwargs.copy()
+                if weights_is_array:
+                    kw["weights"] = weights_to_fit[ii : ii + vox_per_chunk]
+                kwargs_chunks.append(kw)
+
             parallel_kwargs = {}
             for kk in ["n_jobs", "vox_per_chunk", "engine", "verbose"]:
                 if kk in kwargs:
                     parallel_kwargs[kk] = kwargs[kk]
-            fit_array[np.where(mask)] = np.concatenate(
-                (
-                    paramap(
-                        _parallel_fit_worker,
-                        chunks,
-                        func_args=[single_voxel_with_self],
-                        func_kwargs=kwargs,
-                        **parallel_kwargs,
-                    )
-                )
+
+            mvf = paramap(
+                _parallel_fit_worker,
+                chunks,
+                func_args=[single_voxel_with_self],
+                func_kwargs=kwargs_chunks,
+                **parallel_kwargs,
             )
-        return MultiVoxelFit(self, fit_array, mask)
+
+            if isinstance(mvf[0][0], tuple):
+                tmp_fit_array = np.concatenate(
+                    [[svf[0] for svf in mvf_ch] for mvf_ch in mvf]
+                )
+                tmp_extra = np.concatenate(
+                    [[svf[1] for svf in mvf_ch] for mvf_ch in mvf]
+                ).tolist()
+                fit_array[np.where(mask)], extra_list =\
+                    tmp_fit_array, tmp_extra
+                return_extra = True
+            else:
+                tmp_fit_array = np.concatenate(mvf)
+                fit_array[np.where(mask)], extra_list = tmp_fit_array, None
+
+        # Redefine extra to be a single dictionary
+        if return_extra:
+            if extra_list[0] is not None:
+                extra_mask = {key: np.vstack([e[key] for e in extra_list])
+                              for key in extra_list[0]}
+                extra = {}
+                for key in extra_mask:
+                    extra[key] = np.zeros(data.shape)
+                    extra[key][mask == 1] = extra_mask[key]
+            else:
+                extra = None
+
+        # If fit method does not return extra, assume we cannot return extra
+        if return_extra:
+            return MultiVoxelFit(self, fit_array, mask), extra
```

Comment:
I start to wonder if extra should not be part of `Fit` object

Reply:
> The problem is, that DiPy wsa not written in a clever enough way to assume that other things may need returning from the fitting functions, in general.

DIPY is 15years old.... back then, we agree with scikit-learn to follow the same logic... and the need was not the same as today.

and yes, something to consider in the future, always looking for the best solution for the next 15-20years...

Reply:
Happy to help in the future if needs be. It is sort of nice that it's separate from the Fit object, in the sense that it contains info on the fitting but not the results... I find that this isn't a terrible solution. It makes it easy to have the fitting potential return `extra` while leaving the fit objects alone.